nominal/sdk.py

```diff
@@ -0,0 +1,508 @@
+            properties={} if properties is None else dict(properties),
+            start_time=start_abs,
+            title=title,
+            end_time=end_abs,
+        )
+        response = self._run_client.create_run(self._auth_header, request)
+        return Run._from_conjure(self._auth_header, self._run_client, response)
+
+    def get_run(self, run_rid: str) -> Run:
+        """Retrieve a run."""
+        response = self._run_client.get_run(self._auth_header, run_rid)
+        return Run._from_conjure(self._auth_header, self._run_client, response)
+
+    def _list_runs_paginated(self, request: scout_run_api.SearchRunsRequest) -> Iterable[scout_run_api.Run]:
+        while True:
+            response = self._run_client.search_runs(self._auth_header, request)
+            yield from response.results
+            if response.next_page_token is None:
+                break
+            request = scout_run_api.SearchRunsRequest(
+                page_size=request.page_size,
+                query=request.query,
+                sort=request.sort,
+                next_page_token=response.next_page_token,
+            )
+
+    def _list_runs(self) -> Iterable[Run]:
+        # TODO(alkasm): search filters
+        # TODO(alkasm): put in public API when we decide if we only expose search, or search + list.
+        request = scout_run_api.SearchRunsRequest(
+            page_size=100,
+            query=scout_run_api.SearchQuery(),
+            sort=scout_run_api.SortOptions(
+                field=scout_run_api.SortField.START_TIME,
+                is_descending=True,
+            ),
+        )
+        for run in self._list_runs_paginated(request):
+            yield Run._from_conjure(self._auth_header, self._run_client, run)
+
+    def create_dataset_from_io(
+        self,
+        dataset: BinaryIO,
+        name: str,
+        timestamp_column_name: str,
+        timestamp_column_type: TimestampColumnType,
+        file_extension: _AllowedFileExtensions = ".csv",
+        *,
+        description: str | None = None,
+        labels: Sequence[str] = (),
+        properties: Mapping[str, str] | None = None,
+    ) -> str:
+        """Create a dataset from a file-like object.
+        The dataset must be a file-like object in binary mode, e.g. open(path, "rb") or io.BytesIO.
+        If the file is not in binary-mode, the requests library blocks indefinitely.
+
+        Timestamp column types must be a `CustomTimestampFormat` or one of the following literals:
+            "iso_8601": ISO 8601 formatted strings,
+            "epoch_{unit}": epoch timestamps in UTC (floats or ints),
+            "relative_{unit}": relative timestamps (floats or ints),
+            where {unit} is one of: nanoseconds | microseconds | milliseconds | seconds | minutes | hours | days
+        """
+
+        if isinstance(dataset, TextIOBase):
+            raise TypeError(f"dataset {dataset} must be open in binary mode, rather than text mode")
```

Comment:
should we have a version of this method that just takes in a path & opens the file on behalf of the user? helps avoid awkwardness like this too

Reply:
I intended to add that as a separate method in the future; will add it as a todo for now!

Reply:
added as a todo here and n create_attachment